doc: Positional arguments cannot be given by name.

tfiers opened the pull request:
Many new (and old) Julians have been missing this feature over the years.[^1],[^2] But it's not gonna come anytime soon, or ever.[^3]
So, here we add best practices to the docs on how you can still document the names/meaning of positional arguments at the call site.

I added it to the 'Python' section in `noteworthy-differences.md`.
Disadvantage of this location, as [noted][4] by @digital-carver:
- it could also go in the R section.
- individual points on this page (like this one) are not linkable!
- the page in its entirety is already quite long.

The new bullet point _does_ fit in well with the existing, preceding one, however.

[^1]: https://discourse.julialang.org/t/allow-use-of-named-argument-syntax-for-positional-arguments/5287
[^2]: https://discourse.julialang.org/t/naming-positional-arguments-at-call-site/103526
[^3]: https://discourse.julialang.org/t/naming-positional-arguments-at-call-site/103526/97

[4]: https://discourse.julialang.org/t/naming-positional-arguments-at-call-site/103526/112?u=tfiers

tfiers committed:
doc: Positional arguments cannot be given by name. (Python section in `noteworthy-differences.md`)

Many Julians (including me) have been missing this feature over the years.[^1],[^2] But it's not gonna come anytime soon, or ever.[^3]
So, here we add best practices to the docs on how to still document the names/meaning of positional arguments at the call site.

[^1]: https://discourse.julialang.org/t/allow-use-of-named-argument-syntax-for-positional-arguments/5287
[^2]: https://discourse.julialang.org/t/naming-positional-arguments-at-call-site/103526
[^3]: https://discourse.julialang.org/t/naming-positional-arguments-at-call-site/103526/97

tfiers committed:
(make text flow with context)

adienes commented:
personally, I would make it shorter (e.g. stop after "will not work" and omit the remedies) and then just duplicate the note in the R section. I don't think the existence of the ability to name variables is different from Python

tfiers commented:
No it is indeed not different.
The reason we'd wanna add the two workarounds is to point people that miss this feature to them.
(Re what GregVernon and digital_carver said in the [linked thread][1]).
Another good place for this, instead of the "Noteworthy differences" page, might be [the FAQ page][2]. It's a bit more high profile, and it's linkable. Also the multiline code block would not stand out as much as it does here (As it stands, this is the only multiline codeblock in `noteworthy-differences.md`; whereas there are plenty in `faq.md`).

[1]: https://discourse.julialang.org/t/naming-positional-arguments-at-call-site/103526/111
[2]: https://docs.julialang.org/en/v1/manual/faq/

timholy commented:
For completeness, let's xref #51256 (but I think "noteworthy differences" is a good place for this info).

tfiers commented:
@ Tim, thank you, totally missed that one.

@ adienes, on your comment:

Yeah, it might feel odd, "as if it's some kind of trick", whereas it's indeed something very 'stupid'.
But it's in my experience still the best way to do this in Julia.
It's also a relatively highly upvoted [answer][1] in the linked thread.

If I had seen this 'workaround' when learning Julia (in addition to some notice that "This won't be implemented"), I might have spent less time petioning people in Slack/Discourse to implement this in Julia :p

(As to the inline comments method, by the way: I have an (irrational) distaste for it; I never use it).

The other method @baggepinnen mentioned isn't bad either:
```
lognormal(
	4,  # median
	1   # geometric standard deviation, g
)
```

[1]: https://discourse.julialang.org/t/naming-positional-arguments-at-call-site/103526/59?u=tfiers

adienes commented:
> If I had seen this 'workaround' as a Python transplant, (in addition to some notice that "This won't be implemented")

the thing is, I think all that's needed is the "This won't be implemented." and the suggestion to name variables is redundant.

however, I risk repeating myself and am only one vote among many. let's see what others think

brenhinkeller commented on doc/src/manual/noteworthy-differences.md:
````diff
@@ -237,6 +237,16 @@ For users coming to Julia from R, these are some noteworthy differences:
+    median = 4
+    g      = 1  # Geometric standard deviation
+    lognormal(median, g)
+    ```
````
A slightly more compact option might be e.g.
`However, you are free to encode this information in variable names:`
```
median, g = 4, 1
lognormal(median, g)
```
I think without the inline comments example is fine too

GregVernon commented:
> @GregVernon, apologies, totally missed that PR!

No worries, I've cancelled that PR anyways.

tfiers committed:
Make python entry in "Noteworthy differences" shorter. Add, and link to, FAQ section. Add R entry.

tfiers commented:
I shortened the Python entry, added a short entry in the R section, and added (and linked to) a new entry in the FAQ. (Which I argue is a good place for this info, as it indeed seems to be a frequently asked question).

Maybe one thing that could still be added (though not by me, as I'm no expert), is a short explanation of why Julia doesn't allow this.
Though for me, this small PR is already good to merge now, and is hopefully in a state that everyone is happy enough with.

akdienes commented on doc/src/manual/noteworthy-differences.md:
```diff
@@ -158,6 +158,13 @@ For users coming to Julia from R, these are some noteworthy differences:
     will be visible in the caller. This is very different from R and allows new functions to operate
     on large data structures much more efficiently.
+  * In Julia, a function argument is either positional, or a keyword
+    argument. Unlike in R -- where any argument can be passed either
```
probably want three hypens for an em-dash

adienes commented on doc/src/manual/noteworthy-differences.md:
```diff
@@ -158,6 +158,13 @@ For users coming to Julia from R, these are some noteworthy differences:
     on large data structures much more efficiently.
+  * In Julia, a function argument is either positional, or a keyword
+    argument. Unlike in R -- where any argument can be passed either
+    positionally or by name -- in Julia, positional arguments cannot be
```
nit: I would use three hyphens for em-dash and remove surrounding spaces, as well as the second redundant "in Julia," so that would read

>  Unlike in R---where any argument can be passed either positionally or by name---positional arguments...

tfiers replied:
Ah yes great, thank you, will edit.

adienes commented on doc/src/manual/noteworthy-differences.md:
```diff
@@ -158,6 +158,13 @@ For users coming to Julia from R, these are some noteworthy differences:
+    positionally or by name -- in Julia, positional arguments cannot be
+    passed by name, and keyword arguments cannot be passed positionally.
+    E.g. a function with signature `work(x; T)` must be called
+    as `work(4, T = 5)`, and not `work(4, 5)` or `work(x = 4, T = 5)`.
```
nit: probably should recommend `work(4; T = 5)` (note the semicolon there). both formulations function the same, but may as well set the cleanest example

tfiers committed:
implement @adienes 's suggestions

tfiers committed:
fix internal link

DilumAluthge commented:
This PR has had no activity for two years+, so I'll close it as stale.

@tfiers If you're still interested in working on this PR, let me know, and I can re-open it.
